### PR TITLE
Interface call forwarder(icf)

### PR DESCRIPTION
**Pull request by rjnrohit**
- Fixes bug in cairoWriter for contract interface when function input & output arguments spans over multiple lines
- Adds interface forwarder feature
- test for interface call forwarder has been added

### CODE REVIEW

File: src/icf/extInpCheck.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import { assert } from 'console';
```
Comment (GuyMcComb) [inline src/icf/extInpCheck.ts]: Does this assert perform the same task as `import assert = require('assert');`

**Review by GuyMcComb — APPROVED** [review]

File: src/icf/README.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
+## Interface Call Forwarder
```
Comment (swapnilraj) [inline src/icf/README.md:1]: can you add this to the documentation website as well

File: src/icf/README.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+This is a feature for the WARP transpiler to enable interaction between non-warped cairo contracts with a warped cairo contract _(cairo file generated after transpilation of a solidity contract)_.
```
Comment (swapnilraj) [inline src/icf/README.md]: WARP -> Warp

File: src/icf/README.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+```text
+Usage: warp gen_interface [options] <file>
+
+Options:
+  --cairo-path <cairo-path>
+  --output <output>
+  --contract-address <contract-address>
+  --class-hash <class-hash>
```
Comment (swapnilraj) [inline src/icf/README.md]: we should help messages for all options like you have for solc--version

File: src/icf/README.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    --inputs 12 0 13 0
+```
+
+## Running tests
+
+For more detailed and implemented steps, you can look at the [interface_forwarder.test.ts](../../tests/interface_call_forwarder/interface_forwarder.test.ts) file.
+
+To execute interface call forwarder test, run `$ yarn test:forwarder`
+
+## Contact Us :phone:
```
Comment (swapnilraj) [inline src/icf/README.md]: we can remove this section, since its already in the readme

File: src/icf/genCairo.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            acc.pop();
+            acc.push(`${input.name}_len`);
+          }
+          return [...acc, `${input.name}_cast`];
+        }, [])});`;
+
+      const funcDef = (isDelegate: boolean) => [
+        decorator,
+        `func _ITR_${
+          (isDelegate ? '_delegate_' : '') + item.name
```
Comment (swapnilraj) [inline src/icf/genCairo.ts:126]: Why are generating two functions for each cairo function, one with the _delegate_?

File: src/icf/genCairo.ts [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        } {syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr: felt}(`,
+        ...item.inputs.reduce((acc: string[], input: { name: string; type: string }) => {
```
Comment (swapnilraj) [inline src/icf/genCairo.ts]: to improve readability please move these reduce calls outside the array and store them in a variable name

File: src/icf/interfaceCallForwarder.ts [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+ * @returns : none
+ */
+export function generateSolInterface(filePath: string, options: SolcInterfaceGenOptions) {
+  const cairoPathRoot = filePath.slice(0, -'.cairo'.length);
+  const { success } = compileCairo(filePath, path.resolve(__dirname, '..'), {
+    debugInfo: false,
+  });
+
+  // files generated by the compilation of the given cairo contract
+  const resultPath = `${cairoPathRoot}_compiled.json`;
```
Comment (swapnilraj) [inline src/icf/interfaceCallForwarder.ts]: You don't need to "calculate" these yourself, they are returned from compileCairo

File: src/icf/interfaceCallForwarder.ts [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+  if (options.output) {
+    const filePathRoot = options.output.slice(0, -'.sol'.length);
+    solPath = options.output;
+    cairoContractPath = filePathRoot + '.cairo';
+  }
+
+  const abi: AbiType = JSON.parse(fs.readFileSync(abiPath, 'utf8'));
+
+  // generate the cairo contract that will be used to interact with the given cairo contract
+  let cairoContract: string = genCairoContract(abi, options.contractAddress, options.classHash);
```
Comment (swapnilraj) [inline src/icf/interfaceCallForwarder.ts:92]: const cairoContract

File: src/icf/interfaceCallForwarder.ts [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+
+  // replace function names in the generated cairo contract such that it'll match the function
+  // names in the generated solidity contract after transpilation of the latter
+  funcSignatures.forEach((value, key) => {
+    cairoContract = cairoContract.replace(`_ITR_${key}`, `${key}_${value}`);
+  });
+
+  fs.writeFileSync(cairoContractPath, cairoContract);
+
+  const compileForwarder = compileCairo(cairoContractPath, path.resolve(__dirname, '..'), {
```
Comment (swapnilraj) [inline src/icf/interfaceCallForwarder.ts:133]: what are the cases when our generated cairo contract will fail? Can we remove this second compileCairo call?

File: src/icf/utils/struct.ts [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    size: struct.size,
+    members: struct.members.map((member) => {
+      return { name: member.name, offset: member.offset, type: member.type };
+    }),
+  };
+}
+
+export function getStructsFromABI(abi: AbiType): StructAbiItemType[] {
+  const result: StructAbiItemType[] = [];
+  abi.forEach((item) => {
```
Comment (swapnilraj) [inline src/icf/utils/struct.ts]: you can replace this forEach with filter

File: src/icf/utils/struct.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+  abi.forEach((item: AbiItemType) => {
+    if (item.type === 'struct' && item.name !== 'Uint256') {
+      visitStructItemNode(item, visitedStructItem, typeToStruct, result);
+    }
+  });
+  return result;
+}
+
+export function typeToStructMapping(structs: StructAbiItemType[]): Map<string, StructAbiItemType> {
+  const result: Map<string, StructAbiItemType> = new Map();
```
Comment (swapnilraj) [inline src/icf/utils/struct.ts]: new Map(structs.map(s => [s.name,s] ))

File: tests/interface_call_forwarder/contract.cairo [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    warp_external_input_check_int256(a);
+
+    let (__warp_se_0) = warp_add256(a, b);
+
+    return (__warp_se_0,);
+}
+
+@view
+func sub{syscall_ptr: felt*, range_check_ptr: felt, bitwise_ptr: BitwiseBuiltin*}(
+    a: Uint256, b: Uint256
```
Comment (swapnilraj) [inline tests/interface_call_forwarder/contract.cairo]: can you add tests with arrays and felts?

File: tests/interface_call_forwarder/interface_forwarder.test.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+import {
+  declare,
+  deploy,
+  DeployResponse,
+  ensureTestnetContactable,
+  invoke,
+} from '../testnetInterface';
+import * as fs from 'fs';
+
+const testPath = `${path.resolve(__dirname, '..')}/interface_call_forwarder`;
```
Comment (swapnilraj) [inline tests/interface_call_forwarder/interface_forwarder.test.ts]: path.resolve(__dirname, '..', 'interface_call_forwarder');
path.resolve(__dirname, '..', 'interface_call_forwarder', 'contract.cairo');
similar refactoring for all these

File: src/icf/utils/util.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+export function transformType(
+  type: string,
+  typeToStruct: Map<string, StructAbiItemType>,
+  structToAdd?: Map<string, StructAbiItemType>,
+): string {
+  type = type.trim();
+  if (type === 'felt') return 'Uint256';
+  if (typeToStruct.has(type)) return `${type}_uint256`;
+  if (type.endsWith('*')) {
+    return transformType(type.slice(0, -1), typeToStruct, structToAdd) + '*';
```
Comment (swapnilraj) [inline src/icf/utils/util.ts]: nit: template strings

File: src/icf/utils/util.ts [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        .map((_, i) => `${lvar}_${i}`)
+        .join(',')});`,
+    );
+    return castBody.join('\n');
+  }
+  return `${INDENT}let ${lvar} = ${rvar};`;
+}
+
+export function tupleParser(tuple: string): string[] {
+  if (tuple.startsWith('(') && tuple.endsWith(')')) tuple = tuple.slice(1, -1).trim();
```
Comment (swapnilraj) [inline src/icf/utils/util.ts:125]: question: what happens for the type : `((felt, felt), (felt,felt)`

Comment (swapnilraj) [reply src/icf/utils/util.ts:125]: are we not expecting nested tuples, what about static 2d arrays

Comment (rjnrohit) [reply src/icf/utils/util.ts:125]: It'll support nested tuples, line:132 is just to remove the enclosing parenthesis. 
((felt, felt), (felt, felt)) will be split into  [(felt, felt), (felt, felt)]

Comment (rjnrohit) [reply src/icf/utils/util.ts:125]: And, this function doesn't expect faulty tuple type strings. I'll add a case when it receives a wrong string it'll throw an error

File: src/icf/README.md (hunk 2 above)
Comment (rjnrohit) [reply src/icf/README.md:1]: will add it

File: src/icf/interfaceCallForwarder.ts (hunk 9 above)
Comment (rjnrohit) [reply src/icf/interfaceCallForwarder.ts:92]: It's get modified at line : 128

File: src/icf/interfaceCallForwarder.ts (hunk 10 above)
Comment (rjnrohit) [reply src/icf/interfaceCallForwarder.ts:133]: It is needed to do for generation of  class_hash

File: src/icf/utils/struct.ts (hunk 11 above)
Comment (rjnrohit) [reply src/icf/utils/struct.ts]: I was trying `filter` but it was throwing compilation error with message `StructAbiItem[]` is not assignable to `AbiItemType[]`

File: tests/interface_call_forwarder/contract.cairo (hunk 13 above)
Comment (rjnrohit) [reply tests/interface_call_forwarder/contract.cairo]: Adding tests for it

File: docs/docs/features/interface_call_forwarder.mdx [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+/// WARP-GENERATED
+/// class_hash: 0x590f5d5d27ce148a5435e25097968f38b0fe3fe991147a784b1e5c2755c472a
+interface Forwarder_a {
+    function add(uint256 a, uint256 b) external returns (uint256 res);
+}
+```
+
+### Step 2.
+
+Declare and deploy generated cairo file. You can visit here for how to do it: [starknet declare&deploy cairo contracts](https://www.cairo-lang.org/docs/hello_starknet/intro.html#declare-the-contract-on-the-starknet-testnet) .
```
Comment (swapnilraj) [inline docs/docs/features/interface_call_forwarder.mdx]: starknet declare & starknet deploy
missing space and change the wording a bit

File: package.json [hunk 18]
```diff
@@ -28,6 +28,7 @@
     "lint:fix": "npm run lint -- --fix",
     "formatlib": "find warplib/ -iname *.cairo -exec cairo-format -i {} +",
     "test": "yarn warplib && tests/behaviour/setup.sh && npx mocha \"tests/**/*.test.ts\" --extension ts --require ts-node/register --parallel --exit",
+    "test:forwarder": "yarn tsc && npx mocha tests/interface_call_forwarder/interface_forwarder.test.ts --extension ts --require ts-node/register --exit",
```
Comment (swapnilraj) [inline package.json:31]: let's also make sure this test run in CI

Comment (rjnrohit) [reply package.json:31]: CI already covers it, here: https://github.com/NethermindEth/warp/actions/runs/3473030616/jobs/5804584095#step:10:9999

File: src/icf/utils/struct.ts (hunk 11 above)
Comment (swapnilraj) [reply src/icf/utils/struct.ts]: We will have to use type-guards
 https://stackoverflow.com/questions/43010737/way-to-tell-typescript-compiler-array-prototype-filter-removes-certain-types-fro

**Review by swapnilraj — APPROVED** [review]
One minor but apart from the LGTM, thanks for the massive PR

File: src/icf/utils/struct.ts (hunk 11 above)
Comment (rjnrohit) [reply src/icf/utils/struct.ts]: done, thanks!

**Pull request merged**